### PR TITLE
Fix and rename compareTwoFiles method in PackTestZip64

### PR DESCRIPTION
**Pull request by r30shah**
PackTestZip64 was reading jar file into the BufferedInputStream and comparing streams to verify if the jar files created by running pack and unpack command with reference is same or not. This is incorrect way to compare jar files as entries could be laid out differently in the jar file in case the implementation is using hardware support for compressing data. This commit fixes the compare routine (renamed to compareTwoJarFiles) by iterating through individiual entries in jar files and comparing them.

Signed-off-by: Rahil Shah <rahil@ca.ibm.com>

### CODE REVIEW

**Comment by pshipton** [comment]
Verified via https://openj9-jenkins.osuosl.org/view/Test/job/Grinder/1874/

**Comment by r30shah** [comment]
@pshipton / @keithc-ca Given that you have reviewed similar changes in past, can I please get review on this changes?

**Comment by keithc-ca** [comment]
Please first propose the desired change in https://github.com/ibmruntimes/openj9-openjdk-jdk, then back-ports as necessary.

**Comment by r30shah** [comment]
@keithc-ca this test was written for pack200 which are limited to jdk11 and jdk8 lts versions. It wouldn't be there in JDK17 or other actively supported JDK versions or master JDK repo.

**Comment by keithc-ca** [comment]
Perhaps we should follow the practice of the Adoptium project which incorporates the code of conduct by reference and delete it from our repositories. See https://github.com/adoptium and, for example, https://github.com/adoptium/infrastructure/pull/2870.

**Comment by pshipton** [comment]
If the equivalent testing exists in jdk8, we don't run it as part of sanity.openjdk or extended.openjdk afaik, so it's not a concern atm.

File: test/jdk/tools/pack200/PackTestZip64.java [hunk 1]
```diff
@@ -85,43 +85,53 @@ static void testPacking() throws IOException {
         cmdsList.add(unpackedFile.getName());
         Utils.runExec(cmdsList);
 
-        // Compare files binary
-        compareTwoFiles(testFile, unpackedFile);
+        if (!testFile.exists()) {
+            throw new IOException("File " + testFile.getName() + " does not exist!");
+        }
+
+        if(!unpackedFile.exists()) {
```
Comment (keithc-ca) [inline test/jdk/tools/pack200/PackTestZip64.java:92]: Formatting: space before `(`.

File: test/jdk/tools/pack200/PackTestZip64.java [hunk 2]
```diff
@@ -85,43 +85,53 @@ static void testPacking() throws IOException {
...
+            throw new IOException("File " + unpackedFile.getName() + " does not exist!");
+        }
+        // Compare two Jar file
```
Comment (keithc-ca) [inline test/jdk/tools/pack200/PackTestZip64.java:95]: typo: file -> files

File: test/jdk/tools/pack200/PackTestZip64.java [hunk 3]
```diff
@@ -85,43 +85,53 @@ static void testPacking() throws IOException {
...
+                pos++;
+            }
+            srcis.close();
+            dstis.close();
         }
-
-        srcis.close();
-        dstis.close();
+    srcJar.close();
+    dstJar.close();
```
Comment (keithc-ca) [inline test/jdk/tools/pack200/PackTestZip64.java:134]: Issues:
* it is _not_ enough to check for EOF once (the 'dst' entry might be longer)
* don't use raw types
* declare each variable separately
* indent consistently

Suggest:
```java
    static void compareTwoJarFiles(JarFile srcJar, JarFile dstJar) throws IOException {
        Enumeration<JarEntry> entryList = srcJar.entries();
        while (entryList.hasMoreElements()) {
            JarEntry srcEntry = entryList.nextElement();
            JarEntry dstEntry = dstJar.getJarEntry(srcEntry.getName());

            if (dstEntry == null) {
                throw new IOException("Jar Entry " + srcEntry.getName() + " does not exist in " + dstJar.getName());
            }

            BufferedInputStream srcis = new BufferedInputStream(srcJar.getInputStream(srcEntry));
            BufferedInputStream dstis = new BufferedInputStream(dstJar.getInputStream(dstEntry));

            for (int pos = 0;; ++pos) {
                int s = srcis.read();
                int d = dstis.read();

                if (s != d) {
                    throw new IOException("Files are differ starting at position: 0x"
                            + Integer.toHexString(pos));
                } else if (s == -1) {
                	break;
                }
            }

            srcis.close();
            dstis.close();
        }

        srcJar.close();
        dstJar.close();
    }
```

**Review by keithc-ca — CHANGES_REQUESTED** [review]
Please also add the IBM copyright notice.

**Comment by keithc-ca** [comment]
The commit message and the description here should be updated to better describe the intent of this change, which to my mind, is more about correcting the comparison of jar files than the addition of a method.

File: test/jdk/tools/pack200/PackTestZip64.java (hunk 1 above)
Comment (r30shah) [reply test/jdk/tools/pack200/PackTestZip64.java]: Fixed in https://github.com/ibmruntimes/openj9-openjdk-jdk11/compare/3232726537abaceda9f08afd1521668f3f8704c4..18b42c0619ab09f595149f0dacc6c70521482ac8

File: test/jdk/tools/pack200/PackTestZip64.java (hunk 2 above)
Comment (r30shah) [reply test/jdk/tools/pack200/PackTestZip64.java]: Fixed in https://github.com/ibmruntimes/openj9-openjdk-jdk11/compare/3232726537abaceda9f08afd1521668f3f8704c4..18b42c0619ab09f595149f0dacc6c70521482ac8

File: test/jdk/tools/pack200/PackTestZip64.java (hunk 3 above)
Comment (r30shah) [reply test/jdk/tools/pack200/PackTestZip64.java]: Thanks a lot @keithc-ca  for the suggestion. In the initial change for the fix, I reused some of the old code from the existing incorrect compareTwoFiles method. But I agree with your suggestions and have incorporated them in https://github.com/ibmruntimes/openj9-openjdk-jdk11/compare/3232726537abaceda9f08afd1521668f3f8704c4..18b42c0619ab09f595149f0dacc6c70521482ac8

**Comment by r30shah** [comment]
@keithc-ca  I have also changed the commit message and hence the PR description. I do not expect this fundamental change to cause any issue with the test, though will checking it again on internal jenkins (Hence marking it WIP till I have it verified). Would appreciate if you can review again.

File: test/jdk/tools/pack200/PackTestZip64.java [hunk 4]
```diff
@@ -85,43 +91,62 @@ static void testPacking() throws IOException {
...
-        if (!src.exists()) {
-            throw new IOException("File " + src.getName() + " does not exist!");
-        }
-
-        if(!dst.exists()) {
-            throw new IOException("File " + dst.getName() + " does not exist!");
-        }
+    static void compareTwoJarFiles(JarFile srcJar, JarFile dstJar) throws IOException {
+        for (Enumeration srcEntries = srcJar.entries(); srcEntries.hasMoreElements(); ) {
+            JarEntry srcEntry = (JarEntry) srcEntries.nextElement();
```
Comment (keithc-ca) [inline test/jdk/tools/pack200/PackTestZip64.java:112]: Please don't use raw types:
```java
        for (Enumeration<JarEntry> srcEntries = srcJar.entries(); srcEntries.hasMoreElements();) {
            JarEntry srcEntry = srcEntries.nextElement();
```

File: test/jdk/tools/pack200/PackTestZip64.java [hunk 5]
```diff
@@ -85,43 +91,62 @@ static void testPacking() throws IOException {
...
+            JarEntry dstEntry = dstJar.getJarEntry(srcEntry.getName());
+            if (dstEntry == null) {
+                throw new IOException("Jar Entry "+srcEntry.getName()+" does not exist in "+dstJar.getName());
```
Comment (keithc-ca) [inline test/jdk/tools/pack200/PackTestZip64.java:115]: Please insert a space on each side of all binary operators:
```
                throw new IOException("Jar Entry " + srcName + " does not exist in " + dstJar.getName());
```

File: test/jdk/tools/pack200/PackTestZip64.java [hunk 6]
```diff
@@ -85,43 +91,62 @@ static void testPacking() throws IOException {
...
+            }
 
-        BufferedInputStream srcis, dstis;
-        srcis = new BufferedInputStream(new FileInputStream(src));
-        dstis = new BufferedInputStream(new FileInputStream(dst));
+            BufferedInputStream srcis = new BufferedInputStream(srcJar.getInputStream(srcEntry);
+            BufferedInputStream dstis = new BufferedInputStream(dstJar.getInputStream(dstEntry);
```
Comment (keithc-ca) [inline test/jdk/tools/pack200/PackTestZip64.java:119]: Syntax errors; missing `)` on both lines.

File: test/jdk/tools/pack200/PackTestZip64.java [hunk 7]
```diff
@@ -85,43 +91,62 @@ static void testPacking() throws IOException {
...
 
-        int s = 0, d, pos = 0;
-        while (s != -1) { // Checking of just one result for EOF is enough
-            s = srcis.read();
-            d = dstis.read();
+            for (int pos=0; ;++pos) {
```
Comment (keithc-ca) [inline test/jdk/tools/pack200/PackTestZip64.java:121]: Space both side of `=` and after second `;`.

File: test/jdk/tools/pack200/PackTestZip64.java [hunk 8]
```diff
@@ -85,43 +91,62 @@ static void testPacking() throws IOException {
...
 
-            if (s != d) {
-                throw new IOException("Files are differ starting at position: "
-                + Integer.toHexString(pos));
+                if (s != d) {
+                    throw new IOException("Files differ starting at position: "
+                            + Integer.toHexString(pos));
+                }
+                // Byte read from srcis and dstis are same, checking if we reached
+                // end of stream for this jar entry to terminate the loop
```
Comment (keithc-ca) [inline test/jdk/tools/pack200/PackTestZip64.java:130]: Byte -> Bytes, finish sentence with a period.

File: test/jdk/tools/pack200/PackTestZip64.java [hunk 9]
```diff
@@ -85,43 +91,62 @@ static void testPacking() throws IOException {
...
+                }
             }
 
-            pos++;
+            srcis.close();
+            dstis.close();
         }
 
-        srcis.close();
-        dstis.close();
```
Comment (keithc-ca) [inline test/jdk/tools/pack200/PackTestZip64.java:124]: This must be restored (after the loop below).

File: test/jdk/tools/pack200/PackTestZip64.java [hunk 10]
```diff
@@ -85,43 +91,62 @@ static void testPacking() throws IOException {
...
+        // Previous loop will check for all the entries from source jar file and
+        // look for corresponding entry in destination jar to compare. It might
+        // be possible that destination jar file may contain more entries in
+        // which case, the test should fail as well.
+        for (Enumeration dstEntries = dstJar.entries(); dstEntries.hasMoreElements(); ) {
+            JarEntry dstEntry = (JarEntry) dstEntries.nextElement();
+            if (srcJar.getJarEntry(dstEntry.getName()) == null) {
+                throw new IOException("Jar Entry "+dstEntry.getName()+" does not exist in "+srcJar.getName());
```
Comment (keithc-ca) [inline test/jdk/tools/pack200/PackTestZip64.java:147]: Please no raw types; missing spaces around binary operators.

File: test/jdk/tools/pack200/PackTestZip64.java [hunk 11]
```diff
@@ -85,43 +91,62 @@ static void testPacking() throws IOException {
...
-            d = dstis.read();
+            for (int pos=0; ;++pos) {
+                int s = srcis.read();
+                int d = dstis.read();
 
-            if (s != d) {
-                throw new IOException("Files are differ starting at position: "
-                + Integer.toHexString(pos));
+                if (s != d) {
+                    throw new IOException("Files differ starting at position: "
```
Comment (keithc-ca) [inline test/jdk/tools/pack200/PackTestZip64.java:126]: Perhaps you missed it, but I suggested adding "0x" at the end of this string to make it clear the position is expressed in hex. Please add it.

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: test/jdk/tools/pack200/PackTestZip64.java [hunk 12]
```diff
@@ -85,43 +91,62 @@ static void testPacking() throws IOException {
...
+        // Previous loop will check for all the entries from source jar file and
+        // look for corresponding entry in destination jar to compare. It might
+        // be possible that destination jar file may contain more entries in
+        // which case, the test should fail as well.
+        for (Enumeration dstEntries = dstJar.entries(); dstEntries.hasMoreElements(); ) {
+            JarEntry dstEntry = (JarEntry) dstEntries.nextElement();
+            if (srcJar.getJarEntry(dstEntry.getName()) == null) {
+                throw new IOException("Jar Entry "+dstEntry.getName()+" does not exist in "+srcJar.getName());
+            }
+        }
```
Comment (keithc-ca) [inline test/jdk/tools/pack200/PackTestZip64.java:149]: The jar files should be closed:
```java
        srcJar.close();
        dstJar.close();
```

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: test/jdk/tools/pack200/PackTestZip64.java [hunk 13]
```diff
@@ -85,43 +91,64 @@ static void testPacking() throws IOException {
...
-    static void compareTwoFiles(File src, File dst) throws IOException {
-        if (!src.exists()) {
-            throw new IOException("File " + src.getName() + " does not exist!");
-        }
-
-        if(!dst.exists()) {
-            throw new IOException("File " + dst.getName() + " does not exist!");
-        }
+    static void compareTwoJarFiles(JarFile srcJar, JarFile dstJar) throws IOException {
+        for (Enumeration<JarEntry> srcEntries = srcJar.entries(); srcEntries.hasMoreElements(); ) {
```
[the remark this answers is not in the document]
Comment (r30shah) [reply test/jdk/tools/pack200/PackTestZip64.java]: Changed in https://github.com/ibmruntimes/openj9-openjdk-jdk11/compare/c90e742fe22a0d8936317c6974ab336ec14f361f..27ceb89edebee108e1fbd4210050dfd7cabacb8b

File: test/jdk/tools/pack200/PackTestZip64.java (hunk 7 above)
Comment (r30shah) [reply test/jdk/tools/pack200/PackTestZip64.java]: Changed in https://github.com/ibmruntimes/openj9-openjdk-jdk11/compare/c90e742fe22a0d8936317c6974ab336ec14f361f..27ceb89edebee108e1fbd4210050dfd7cabacb8b

File: test/jdk/tools/pack200/PackTestZip64.java (hunk 6 above)
Comment (r30shah) [reply test/jdk/tools/pack200/PackTestZip64.java]: Fixed in https://github.com/ibmruntimes/openj9-openjdk-jdk11/compare/a16261f995e7aaf04f90db78cbaf9188c987c6a9..c90e742fe22a0d8936317c6974ab336ec14f361f

File: test/jdk/tools/pack200/PackTestZip64.java (hunk 5 above)
Comment (r30shah) [reply test/jdk/tools/pack200/PackTestZip64.java]: Fixed in https://github.com/ibmruntimes/openj9-openjdk-jdk11/compare/a16261f995e7aaf04f90db78cbaf9188c987c6a9..c90e742fe22a0d8936317c6974ab336ec14f361f

File: test/jdk/tools/pack200/PackTestZip64.java (hunk 4 above)
Comment (r30shah) [reply test/jdk/tools/pack200/PackTestZip64.java]: Fixed in https://github.com/ibmruntimes/openj9-openjdk-jdk11/compare/a16261f995e7aaf04f90db78cbaf9188c987c6a9..c90e742fe22a0d8936317c6974ab336ec14f361f

**Comment by keithc-ca** [comment]
Please launch a grinder like the one mentioned in https://github.com/ibmruntimes/openj9-openjdk-jdk11/pull/615#issuecomment-1402799119 and share the link here.

**Comment by r30shah** [comment]
I can not launch grinder on OpenJ9 jenkins, though I have launched one in internal jenkins (`/job/Grinder/30737/console`)

**Comment by pshipton** [comment]
https://openj9-jenkins.osuosl.org/view/Test/job/Grinder/1895/

**Comment by pshipton** [comment]
Previous grinder passed. If there are more changes, it can just be restarted.

**Comment by r30shah** [comment]
@keithc-ca Grinders from @pshipton  in https://openj9-jenkins.osuosl.org/view/Test/job/Grinder/1895/ passed and so as the one I launched in internal jenkins. If you are good with changes, can we merge this one?

**Review by keithc-ca — APPROVED** [review]

**Pull request merged**